src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs

```diff
@@ -193,20 +197,44 @@ public override Task AnalyzeProjectAsync(Project project, bool semanticsChanged,
                 return UpdateSymbolTreeInfoAsync(project, cancellationToken);
             }
 
+            [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/33172", AllowCaptures = false)]
             private async Task UpdateSymbolTreeInfoAsync(Project project, CancellationToken cancellationToken)
             {
                 Debug.Assert(SupportAnalysis(project));
 
                 // Produce the indices for the source and metadata symbols in parallel.
-                var tasks = new List<Task>
-                {
-                    GetTask(project, () => UpdateSourceSymbolTreeInfoAsync(project, cancellationToken), cancellationToken),
-                    GetTask(project, () => UpdateReferencesAync(project, cancellationToken), cancellationToken)
-                };
+                var isRemoteWorkspace = project.Solution.Workspace.Kind == WorkspaceKind.RemoteWorkspace;
+                var (sourceSymbolTask, referencesTask) = isRemoteWorkspace
+                    ? UpdateSymbolTreeInfoInServiceProcessAsync(project, cancellationToken)
+                    : UpdateSymbolTreeInfoInCurrentProcessAsync(project, cancellationToken);
+
+                var tasks = s_taskArrayPool.Allocate();
+                tasks[0] = sourceSymbolTask;
+                tasks[1] = referencesTask;
 
-                await Task.WhenAll(tasks).ConfigureAwait(false);
+                await Task.WhenAll(tasks)
+                    .ContinueWith((t, state) =>
+                    {
+                        var tasks = (Task[])state;
+
+                        tasks[0] = tasks[1] = null;
+                        s_taskArrayPool.Free(tasks);
+                    },
+                    tasks,
+                    cancellationToken,
+                    TaskContinuationOptions.ExecuteSynchronously,
+                    TaskScheduler.Default).ConfigureAwait(false);
```

Comment:
do you need the await/COnfigureAwait?  Can you not just return the `Task.WhenAll`? Then, this function won't need to be 'async' either right?

Reply:
I could, but I'm usually wary of removing `await` as it changes exception handling behaviour. But since most callers do end up using `await` somewhere, and this isn't in a try/catch, it should be okay.

Reply:
I'll make that change.

Reply:
understand the concern.  though in roslyn, unless yo'ure the top-level kick-off point, it should be  fine.

Reply:
Yeah, I'll change it :+1: